pandas/core/ops/array_ops.py

```diff
@@ -132,3 +153,233 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
+
+    rvalues = maybe_upcast_for_op(rvalues, lvalues.shape)
+
+    if should_extension_dispatch(left, rvalues) or isinstance(
+        rvalues, (ABCTimedeltaArray, ABCDatetimeArray, Timestamp)
+    ):
+        # TimedeltaArray, DatetimeArray, and Timestamp are included here
+        #  because they have `freq` attribute which is handled correctly
+        #  by dispatch_to_extension_op.
+        res_values = dispatch_to_extension_op(op, lvalues, rvalues, keep_null_freq)
+
+    else:
+        with np.errstate(all="ignore"):
+            res_values = na_arithmetic_op(lvalues, rvalues, op, str_rep, eval_kwargs)
+
+    return res_values
+
+
+def comparison_op(
+    left: Union[np.ndarray, ABCExtensionArray], right: Any, op
+) -> Union[np.ndarray, ABCExtensionArray]:
+    """
+    Evaluate a comparison operation `=`, `!=`, `>=`, `>`, `<=`, or `<`.
+
+    Parameters
+    ----------
+    left : np.ndarray or ExtensionArray
+    right : object
+        Cannot be a DataFrame, Series, or Index.
+    op : {operator.eq, operator.ne, operator.gt, operator.ge, operator.lt, operator.le}
+
+    Returns
+    -------
+    ndarrray or ExtensionArray
+    """
+    from pandas.core.ops import should_extension_dispatch, dispatch_to_extension_op
+
+    # NB: We assume extract_array has already been called on left and right
+    lvalues = left
+    rvalues = right
+
+    rvalues = lib.item_from_zerodim(rvalues)
+    if isinstance(rvalues, list):
+        # TODO: same for tuples?
+        rvalues = np.asarray(rvalues)
+
+    if isinstance(rvalues, (np.ndarray, ABCExtensionArray, ABCIndexClass)):
+        # TODO: make this treatment consistent across ops and classes.
+        #  We are not catching all listlikes here (e.g. frozenset, tuple)
+        #  The ambiguous case is object-dtype.  See GH#27803
+        if len(lvalues) != len(rvalues):
+            raise ValueError("Lengths must match to compare")
+
+    if should_extension_dispatch(lvalues, rvalues):
+        res_values = dispatch_to_extension_op(op, lvalues, rvalues)
+
+    elif is_scalar(rvalues) and isna(rvalues):
+        # numpy does not like comparisons vs None
+        if op is operator.ne:
+            res_values = np.ones(len(lvalues), dtype=bool)
+        else:
+            res_values = np.zeros(len(lvalues), dtype=bool)
+
+    elif is_object_dtype(lvalues.dtype):
+        res_values = comp_method_OBJECT_ARRAY(op, lvalues, rvalues)
+
+    else:
+        op_name = "__{op}__".format(op=op.__name__)
+        method = getattr(lvalues, op_name)
+        with np.errstate(all="ignore"):
+            res_values = method(rvalues)
+
+        if res_values is NotImplemented:
+            res_values = invalid_comparison(lvalues, rvalues, op)
+        if is_scalar(res_values):
+            raise TypeError(
+                "Could not compare {typ} type with Series".format(typ=type(rvalues))
+            )
+
+    return res_values
+
+
+def na_logical_op(x, y, op):
+    try:
```

Comment:
if you can type / doc-string at some point

Reply:
Sure.  This function is headed for a pretty major re-write in follow-ups.